src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs

```diff
@@ -58,7 +58,47 @@ public class TemplateBinder
         // Step 1: Get the list of values we're going to try to use to match and generate this URI
         public TemplateValuesResult GetValues(RouteValueDictionary ambientValues, RouteValueDictionary values)
         {
-            var context = new TemplateBindingContext(_defaults);
+            // Any default values that don't appear as parameters are treated like filters. Any new values
+            // provided must match these defaults.
+            foreach (var filter in _filters)
+            {
+                var parameter = GetParameter(filter.Key);
+                if (parameter != null)
+                {
+                    continue;
+                }
+
+                object value;
+                if (values.TryGetValue(filter.Key, out value))
+                {
+                    if (!RoutePartsEqual(value, filter.Value))
+                    {
+                        // If there is a non-parameterized value in the route and there is a
+                        // new value for it and it doesn't match, this route won't match.
+                        return null;
+                    }
+                }
+            }
+
+            // Validate that all required parameters have a value.
+            // Perf: Fail fast if the required parameter doesn't have a value.
+            for (var i = 0; i < _template.Parameters.Count; i++)
+            {
+                var parameter = _template.Parameters[i];
+                if (parameter.IsOptional || parameter.IsCatchAll)
+                {
+                    continue;
+                }
+
+                if (!HasValue(parameter.Name, ambientValues, values))
```

Comment:
This check is not correct because you have to take into account route-value-invalidation. Not all ambient values are eligible to be used in link generation.

Basically I would expect your changes to start generating links in cases where routing would not normally be able to.

Consider this:
```
route: {controller}/{action}/{id?}
ambient values: new { controller = "Home", action = "Index" }
values: new { controller = "Blog" }
```

In this case, the value for `controller` has changed, which means that any ambient values that appear lexically-after `controller` in the template should not be used, so the value of `action` in the ambient values cannot be used. This is handled [here](https://github.com/aspnet/Routing/pull/314/files#diff-46626dd8adef7c6d9adc039d27a7201dL85) in the existing code.